File: algorithms/javaalgo/src/test/java/programmers/level2/englishfollowup/EnglishFollowUpTest.java
package programmers.level2.englishfollowup;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.HashSet;
import java.util.Set;
import org.junit.jupiter.api.Test;

public class EnglishFollowUpTest {

    @Test
    void test1() {
        String[] words = {"tank", "kick", "know", "wheel", "land", "dream", "mother", "robot", "tank"};

        assertThat(solution(3, words)).isEqualTo(new int[]{3, 3});
    }

    @Test
    void test2() {
        String[] words = {"hello", "one", "even", "never", "now", "world", "draw"};
        assertThat(solution(2, words)).isEqualTo(new int[]{1, 3});
    }

    public int[] solution(int n, String[] words) {
        int[] answer = new int[2];
        Set<String> wordSet = new HashSet<>();
        wordSet.add(words[0]);
        for (int i = 1; i < words.length; i++) {
            String word = words[i];
            String beforeWord = words[i - 1];
            if (!wordSet.contains(word) && beforeWord.charAt(beforeWord.length() - 1) == word.charAt(0)) {
                wordSet.add(word);
            } else {
                answer[0] = (i % n) + 1;
                answer[1] = (i / n) + 1;
                return answer;
            }
        }
        return answer;
    }
}

File: algorithms/javascriptalgo/programmers/level2/englishFollowUp.test.js
test("test", () => {
    let n1 = 3;
    let words1 = ["tank", "kick", "know", "wheel", "land", "dream", "mother", "robot", "tank"];

    let n2 = 2;
    let words2 = ["hello", "one", "even", "never", "now", "world", "draw"];

    expect(solution(n1, words1)).toEqual([3, 3]);
    expect(solution(n2, words2)).toEqual([1, 3]);
});

function solution(n, words) {
    let answer = [0, 0];
    const wordSet = new Set();
    wordSet.add(words[0]);

    for (let i = 1; i < words.length; i++) {
        let word = words[i];
        let beforeWord = words[i - 1];
        if (!wordSet.has(word) && beforeWord[beforeWord.length - 1] === word[0]) {
            wordSet.add(word);
        } else {
            answer[0] = (i % n) + 1;
            answer[1] = Math.floor((i / n)) + 1;
            return answer;
        }
    }
    return answer;
}
